File: Maxcut Code/MaxcutRandom.m
%Set up a random graph
n = 50;

p = 0.4;
A = rand(n) < p;
A = triu(A) + triu(A,1)';
A = A - diag(diag(A));

cut = gw_MaxCut(A, 100);
disp(cut)

function cut = gw_MaxCut(A, T)
    [n,~] = size(A);
    % Use CVX
    cvx_begin quiet
        variable X(n,n) symmetric
        minimize trace(A*X)
            diag(X) == ones(n,1);
            X == semidefinite(n);
    cvx_end

    %%
    U = chol(X);
    
    cut = 0;
    for i = 1:T
        r = mvnrnd(zeros(n,1),diag(ones(n,1)))';
        y = sign(U*r);
        cut = cut + (sum(A(:)) - y'*A*y)/4;
    end
    cut = round(cut / T);
end

File: Maxcut Code/MaxcutRandom.py
import cvxpy as cp
import numpy as np

def cut_size(A, T) :
    # Define and solve the CVXPY problem.
    # Create a symmetric matrix variable and constraints
    n = A.shape[0]

    X = cp.Variable((n,n), symmetric=True)
    constraints = [X >> 0]
    constraints += [ cp.diag(X) == 1]

    prob = cp.Problem(cp.Minimize(cp.trace(A @ X)), constraints)
    prob.solve(solver='CVXOPT')

    # Solve for the maximum cut
    U = np.linalg.cholesky(X.value)

    cut = 0
    for i in range(0, T) :
        r = np.random.normal(0, 1, n)
        y = np.sign(U @ r)

        # Calculate the cut
        cut = cut + (np.sum(A) - y.T@A@y)/4
    return round(cut / T)

# Generate a random SDP.
n = 20
p = 0.4
A = (np.random.rand(n, n) < p).astype(int)
A = np.triu(A) + np.triu(A).T
A = A - np.diag(np.diag(A))

cut = cut_size(A, 100)

print(cut)

